Fix BFD support on OpenBSD

shindere opened the pull request:
This is a follow-up to both #8947 and #8749.

The code proposed in #8749 does not work on the OpenBSD machine where I
tried it, because `bfd.h` is installed in `/usr/local/include` rather
than `/usr/include` and also because `-ldl` is in the way (the
library is not found, making all the tests fail, and it is anyway not
necessary).

So this PR fixes all this and makes sure the BFD support gets tested
on Inria's continuous integration infrastructure.

It would be nice to have this in 4.10, @Octachron

shindere committed:
Fix BFD support on OpenBSD

shindere commented:
(The Changes file has been updated assuming the change will be in 4.10)

gasche commented:
What is the amount of testing that your change received? (In particular, on other BSD systems that OpenBSD?).

There was some pretty arcane code in the part removed by your patch ("Override any GCC internal prototype to avoid ..."), and it's not clear from your description why this complicated logic is not necessary anymore. (Your description suggests that we need to try more things to also support OpenBSD, but it doesn't explain why we can remove stuff.) Can you explain what it was doing and why it is it safe to remove it?

shindere commented:
Gabriel Scherer (2019/11/22 07:02 -0800):
> What is the amount of testing that your change received? (In
> particular, on other BSD systems that OpenBSD?).

I tried on FreeBSD but the `bfd.h` that I got installed there did not
compile, I don't know why.

> There was some pretty arcane code in the part removed by your patch
> ("Override any GCC internal prototype to avoid ..."), and it's not
> clear from your description why this complicated logic is not
> necessary anymore. (Your description suggests that we need to try more
> things to also support OpenBSD, but it doesn't explain why we can
> remove stuff.) Can you explain what it was doing and why it is it safe
> to remove it?

I'm not sure what you are referring to. The parts I removed where those
using `-ldl` and I think I explained that this was not necessary and
causing errors. I believe that's the only thing I removed, isn't it?

dbuenzli commented:
That's a bit of a side question but it is somehow related. 

As far as I know (but that's not much), bfd is only used by `objinfo_helper.c` in order to read OCaml runtime system information about `cmxs` file. 

The question is why this is not simply done by `Dynlinking` the `cmxs` file in `ocamlobjinfo` in order to access this information. See [here](https://gist.github.com/dbuenzli/0b773f4a4dd9f7a35f30cd9b671e48c5) for a prototype. 

Is there a bootstrap issue with `ocamlobjinfo` using `Dynlink` ? One problem I can see is a discrepancy between the bytecode and native version of the tool, but can't we cope with this ?

shindere commented:
I have no clue Daniel, sorry. Could it be that the current approach is
able to deal with `.cmxs` of different OCaml versions, which your
approach would not deal with?

damiendoligez commented on configure.ac:
```diff
@@ -1505,17 +1510,18 @@ AS_IF([test x"$with_bfd" != "xno"],
-    AS_IF([test -z "$bfd_ldlibs"],
-      [AC_CHECK_LIB([bfd], [bfd_openr],
-        [bfd_ldlibs="-lbfd -ldl -liberty"], [], [-ldl -liberty])])
+      [unset ac_cv_lib_bfd_bfd_openr
```
I don't understand why you need to unset this variable. The previous code managed to try several alternatives without unsetting anything, right?

gasche commented:
The comment line I referred to was part of the generated `configure`. I suppose that the reason it's not produced anymore comes from the new `unset ac_cv_lib_bfd_bfd_openr` lines you added in configure.ac, which I suppose change something in the way `openr` (whatever that is) is detected. (I see that the complex code generated in `configure` does use `-ldl`, so maybe that's why you got rid of it, but is it the right change?)

@dbuenzli if you were willing to turn your approach into a pull request against tools/objinfo.ml, we could have a discussion of it there (it would be nice in such a PR of properly handling `ndl_open` failure cases, turning them into a proper user message). It looks pleasingly simpler to me, but I don't understand much in this problem space. (I would assume that natdynlink was implemented after ocamlobjinfo, and people never considered your question before.) Naive questions I have (not being familiar with natdynlink) is whether your approach runs the risk of running code from the .cmxs file (we don't want that), or would fail if its dependencies are not present in the current executable (this also sounds bad); I'm happy to keep there for a separate issue/PR.

damiendoligez reviewed (commented):
I have only one question.

madroach commented on configure:
```diff
@@ -15989,6 +15989,13 @@ if test x"$with_bfd" != "xno"; then :
+  BFD_INCLUDE_DIR="/usr/local/include"
+fi
+      if test -z "$BFD_LIB_DIR"; then :
+  BFD_LIB_DIR="/usr/local/lib"
```
I believe this is not such a good idea. I don't think this needs a fixing _at all_.
OpenBSD has all ports installed in ``/usr/local``. So ``OpenBSD`` should provide this information. Either by setting ``BFD_LIB_DIR`` / ``BFD_INCUDE_DIR`` or ``LIBRARY_PATH`` / ``CPATH`` for us.
I would strongly argue against configuring something like this conditional on the OS.
If you really think the configure script should discover the bfd library without outside help you may search for it in common places (``/usr/``, ``/usr/local``, ``/opt``). But as I said I don't think this needs fixing in the OCaml build system. I am regularly maintaining the OCaml port on OpenBSD. It already has 
``` make
CFLAGS +=	-I${LOCALBASE}/include
LDFLAGS +=	-L${LOCALBASE}/lib
```
in it. Everyone else should set ``LIBRARY_PATH`` and ``CPATH`` when building projects supposed to find stuff in ``/usr/local``.

dbuenzli replied:
Note that somehow what @madroach mentions is the same conclusion we eventually arrived to thanks to @ygrek in `opam`, see [here for a discussion](https://github.com/ocaml/opam-repository/issues/3093#issuecomment-312929998).

shindere commented:
Damien Doligez (2019/11/22 15:49 +0000):
> damiendoligez commented on this pull request.
>
> I have only one question.
>
> > @@ -1505,17 +1510,18 @@ AS_IF([test x"$with_bfd" != "xno"],
>      [bfd_ldlibs=""
>      AC_CHECK_LIB([bfd], [bfd_openr], [bfd_ldlibs="-lbfd"])
>      AS_IF([test -z "$bfd_ldlibs"],
> -      [AC_CHECK_LIB([bfd], [bfd_openr], [bfd_ldlibs="-lbfd -ldl"], [], [-ldl])])
> -    AS_IF([test -z "$bfd_ldlibs"],
> -      [AC_CHECK_LIB([bfd], [bfd_openr],
> -        [bfd_ldlibs="-lbfd -ldl -liberty"], [], [-ldl -liberty])])
> +      [unset ac_cv_lib_bfd_bfd_openr
>
> I don't understand why you need to unset this variable. The previous
> code managed to try several alternatives without unsetting anything,
> right?

No, it failed to do so because the subsequent calls to the macro sued
the cached value and did not really do the test, as has been explained
in #8749 wher e you yourself commented. ;-)

Unsetting the variable is actually essential.

shindere commented:
Gabriel Scherer (2019/11/22 07:49 -0800):
> The comment line I referred to was part of the generated `configure`.
> I suppose that the reason it's not produced anymore comes from the new
> `unset ac_cv_lib_bfd_bfd_openr` lines you added in configure.ac, which
> I suppose change something in the way `openr` (whatever that is) is
> detected.

It's about results cached by autoconf.

> (I see that the complex code generated in `configure` does use `-ldl`,
> so maybe that's why you got rid of it, but is it the right change?)

Well reviewing generated code makes the discussion a bit complex,
IMO. ;-)

dbuenzli commented:
> @dbuenzli if you were willing to turn your approach into a pull request against tools/objinfo.ml, we could have a discussion of it there (it would be nice in such a PR of properly handling `ndl_open` failure cases, turning them into a proper user message)

I'm afraid that won't happen *soon*. I was rather fishing for an obvious rebutal to the approach from someone familiar with the topic. I can open an issue if people think it's worth it.

However: 

> Naive questions I have (not being familiar with natdynlink) is whether your approach runs the risk of running code from the .cmxs file (we don't want that)

I don't think this should happen. Unless I'm mistaken [`caml_natdynlink_open`](https://github.com/ocaml/ocaml/blob/d3e73595e55e84250fa77f04e9c239dee1224b7b/runtime/dynlink_nat.c#L66-L96) simply loads the code, which should not execute anything. There's a bit of churn done after the load in order to initialize the modules (I believe the relevant part is [there](https://github.com/ocaml/ocaml/blob/fe31a0179dd0575f217dcde11ca4dabcc2673c57/otherlibs/dynlink/dynlink_common.ml#L333-L352)).

> or would fail if its dependencies are not present in the current executable (this also sounds bad); I'm happy to keep there for a separate issue/PR.

Yes that's part of the questions. E.g. what happens with the `caml_natdynlink_open` when dynlink is not supported (for the approach to work we should make it available but raise) so it needs a closer look. Unfortunately busy right now®.

dra27 commented:
I agree with @madroach’s assessment that it’s not nice to be hard-coding these alternative search paths for different platforms, although a better compromise might be to allow a fallback to `/usr/local` for _all_ platforms?

Equally, all hacks to one side - isn’t this exactly what `libtool` is supposed to deal with, and might we do better to start using it here? It’s needed in one place (and all 6 Windows ports can’t use it anyway - at least on Cygwin it didn’t work when I last checked)

shindere commented:
madroach (2019/11/22 07:53 -0800):
> madroach commented on this pull request.
>
>
>
> > @@ -15989,6 +15989,13 @@ if test x"$with_bfd" != "xno"; then :
>  fi
>        if test -z "$BFD_LIB_DIR"; then :
>    BFD_LIB_DIR="/opt/local/lib"
> +fi ;; #(
> +  *-*openbsd*) :
> +    if test -z "$BFD_INCLUDE_DIR"; then :
> +  BFD_INCLUDE_DIR="/usr/local/include"
> +fi
> +      if test -z "$BFD_LIB_DIR"; then :
> +  BFD_LIB_DIR="/usr/local/lib"

It may make things easier to comment on the source autoconf rather than
on the generated shell script.

> I believe this is not such a good idea. I don't think this needs a fixing _at all_.
> OpenBSD has all ports installed in ``/usr/local``. So ``OpenBSD``
> should provide this information. Either by setting ``BFD_LIB_DIR`` /
> ``BFD_INCUDE_DIR`` or ``LIBRARY_PATH`` / ``CPATH`` for us.

[...]

Just to make sure: did you understand that what configure does is jsut
providing *defaults* that can be overriden by the user?

shindere commented:
David Allsopp (2019/11/22 09:29 -0800):
> I agree with @madroach’s assessment that it’s not nice to be
> hard-coding these alternative search paths for different platforms,
> although a better compromise might be to allow a fallback to
> `/usr/local` for _all_ platforms?

As I wrote, the only thing that is hardcoded is a default. The
possibility to override it remains and user has the last word.

> Equally, all hacks to one side - isn’t this exactly what `libtool` is
> supposed to deal with, and might we do better to start using it here?
> It’s needed in one place (and all 6 Windows ports can’t use it anyway
> - at least on Cygwin it didn’t work when I last checked)

No it's not libtool's job. libtool is for making it easier to build
libraries and using them both from the build tree and once installed. I
think you meant pkg-config, right?

shindere commented:
In order to (hopefully) come to an agreement:

 * I believe this PR is a strict improvmenet over the current situation

 * It only provides defaults that can be overriden by the user

 * Providing such OS-dependent defaults was already done for MacOS

It is certainly possible to search the three directories for header
and library files on all operating systems, but that would require
either t do a bit of copy/paste, or to write an M4 macro to search a list of
directories, if one wants to keep the code factorised.

To be honest I have neither the time nor the motivation to write such a
macro right now and am not very excited by the code duplication either.

So I would like to suggest that the PR gets merged in its present state
and that the potential improvments are left as future work.

dra27 commented:
@shindere - no, I meant `libtool` (`pkg-config` is a related tool) - IIUC, you provide the paths to `configure` (`/usr/local/lib` on OpenBSD, it appears `/usr/pkg/lib` on NetBSD, from my poking) and then use `libtool` to compile the helper program - for example, on the NetBSD VM I just fired up, `libbfd.la` tells `libtool` that you only need `-lz` to link with libbfd. The point is that it removes the need to detect the linker flags, which should eliminate a lot autoconf code. Also, at least for a GNU project like binutils, I think you can have reasonable hope of `libbfd.la` being installed, where `pkg-config`'s `.pc` files are not guaranteed to be available anywhere.

I agree that the present of this PR is an improvement, however.

shindere commented:
Thanks, @Êra27. So that means we would use `libtool` forlinking but
still have to look for the header files the way we are doing right now,
right?

To be honest I'd rather leave that for a later PR which could e.g. be
more specifically dedicated to using libtool. What do you think?

dra27 commented:
That's my understanding, yes. I agree with leaving the change to a future PR and taking this as-is for now.

shindere commented:
Thanks!

shindere commented:
Thanks again, @dra27. I merged and also cherry-picked to the 4.10 branch
as commit 109358ecf8bf552f0e5bd37de435dfe25028176c
